src/main/java/ca/teamdman/sfm/common/capabilityprovidermapper/ae2/InterfaceCapabilityProviderMapper.java

```diff
@@ -0,0 +1,446 @@
+            if (maybeCap.isPresent()) {
+                //noinspection DataFlowIssue
+                return callback.apply(maybeCap.orElse(null));
+            }
+            return null;
+        }
+
+        @Override
+        public int getSlots() {
+            Integer slots = this.withStorage(s -> {
+                int i = 0;
+                for (var stored : s.getAvailableStacks()) {
+                    if (stored.getKey() instanceof AEItemKey) {
+                        i++;
+                    }
+                }
+                return i;
+            });
+
+            if (slots == null) {
+                slots = this.withBaseItemHandler(IItemHandler::getSlots);
+            }
+
+            return slots == null ? 0 : slots;
+        }
+
+        @Override
+        public @NotNull ItemStack getStackInSlot(int slot) {
+            var stack = this.withStorage(s -> {
+                int i = 0;
+                for (var stored : s.getAvailableStacks()) {
+                    if (stored.getKey() instanceof AEItemKey key) {
+                        if (slot == i++) {
+                            return key.toStack((int) Math.min(Integer.MAX_VALUE, stored.getLongValue()));
+                        }
+                    }
+                }
+                return ItemStack.EMPTY;
+            });
+
+            if (stack == null) {
+                stack = this.withBaseItemHandler(c -> c.getStackInSlot(slot));
+            }
+
+            return stack == null ? ItemStack.EMPTY : stack;
+        }
+
+        @Override
+        public @NotNull ItemStack insertItem(int slot, @NotNull ItemStack stack, boolean simulate) {
+            if (stack.isEmpty()) {
+                return stack;
+            }
+
+            Integer inserted = this.withStorage(s -> {
+                var key = AEItemKey.of(stack);
+                if (key == null) {
+                    return 0;
+                }
+
+                var energy = this.getEnergy();
+                if (energy == null) {
+                    return 0;
+                }
+
+                return (int) StorageHelper.poweredInsert(
+                        energy,
+                        s,
+                        key,
+                        stack.getCount(),
+                        IActionSource.empty(),
+                        simulate ? Actionable.SIMULATE : Actionable.MODULATE
+                );
+            });
+
+            if (inserted == null) {
+                var stack2 = this.withBaseItemHandler(c -> c.insertItem(slot, stack, simulate));
+                return stack2 == null ? ItemStack.EMPTY : stack2;
+            }
+
+            if (!simulate) {
+                stack.shrink(inserted);
+                return stack;
+            }
+
+            var rtn = stack.copy();
+            rtn.shrink(inserted);
+            return rtn;
+        }
+
+        @Override
+        public @NotNull ItemStack extractItem(int slot, int amount, boolean simulate) {
```

Comment:
```suggestion
        public ItemStack extractItem(int slot, int amount, boolean simulate) {
```
the package file should be adding notnullbydefault to everything